From 9e038e0883292bbf7b24e9f0c145e4aa8fd29740 Mon Sep 17 00:00:00 2001
From: fepitre <frederic.epitre@orange.fr>
Date: Sun, 13 Aug 2017 10:56:47 +0200
Subject: [PATCH] pvops: blkfront removable flag

Source: QubesOS

Signed-off-by: fepitre <frederic.epitre@orange.fr>
---
 drivers/block/xen-blkfront.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/block/xen-blkfront.c b/drivers/block/xen-blkfront.c
index 3945963..3591be0 100644
--- a/drivers/block/xen-blkfront.c
+++ b/drivers/block/xen-blkfront.c
@@ -2361,6 +2361,7 @@ static void blkfront_connect(struct blkfront_info *info)
 	unsigned int binfo;
 	char *envp[] = { "RESIZE=1", NULL };
 	int err, i;
+	int removable;
 
 	switch (info->connected) {
 	case BLKIF_STATE_CONNECTED:
@@ -2428,6 +2429,12 @@ static void blkfront_connect(struct blkfront_info *info)
 		}
 	}
 
+	err = xenbus_gather(XBT_NIL, info->xbdev->otherend,
+			    "removable", "%d", &removable,
+			    NULL);
+	if (!err && removable)
+		binfo |= VDISK_REMOVABLE;
+
 	err = xlvbd_alloc_gendisk(sectors, info, binfo, sector_size,
 				  physical_sector_size);
 	if (err) {
-- 
2.5.5

